src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -0,0 +1,728 @@
+            and isinstance(cols[0], UnresolvedAlias)
+            and isinstance(cols[0].child, Star)
+        ):
+            return self
+        try:
+            final_projection = []
+            new_column_states = derive_column_states_from_subquery(cols, self)
+            disable_next_level_flatten = False
+            if self.flatten_disabled or self._has_clause_using_columns():
+                can_flatten = False
+            else:
+                can_flatten = True
+                subquery_column_states = self.column_states
+                for col, state in new_column_states.items():
+                    dependent_columns = state.dependent_columns
+                    if dependent_columns == COLUMN_DEPENDENCY_DOLLAR:
+                        can_flatten = False
+                        break
+                    subquery_state = subquery_column_states.get(col)
+                    if state.change_state in (
+                        ColumnChangeState.CHANGED_EXP,
+                        ColumnChangeState.NEW,
+                    ):
+                        can_flatten = can_projection_dependent_columns_flatten(
+                            dependent_columns, subquery_column_states
+                        )
+                        if not can_flatten:
+                            break
+                        final_projection.append(state.expression)
+                    elif state.change_state == ColumnChangeState.UNCHANGED_EXP:
+                        # query may change sequence of columns. If subquery has same-level reference, flattened sql may not work.
+                        if subquery_column_states[col].depend_on_same_level:
+                            can_flatten = False
+                            break
+                        final_projection.append(
+                            subquery_column_states[col].expression
+                        )  # add subquery's expression for this column name
+                    else:  # state == ColumnChangeState.DROPPED:
+                        if (
+                            subquery_state.change_state == ColumnChangeState.NEW
+                            and subquery_state.referenced_by_same_level_column
+                        ):
+                            can_flatten = False
+                            break
+        except (ParseColumnNameError, DeriveColumnDependencyError):
+            can_flatten = False
+            disable_next_level_flatten = True
+            final_projection = cols
+
+        if can_flatten:
+            new = copy(self)
+            new.projection_ = final_projection
+            new.from_ = self.from_.to_subqueryable()
+            new.pre_actions = new.from_.pre_actions
+            new.post_actions = new.from_.post_actions
+        else:
+            new = SelectStatement(
+                projection_=cols, from_=self.to_subqueryable(), analyzer=self.analyzer
+            )
+            new.flatten_disabled = disable_next_level_flatten
```

Comment:
Even though the end result is the same for the current implementation, this probably should go outside of `else`? Whether the new one can be flattened is not directly related to `can_flatten`.